### PR TITLE
feat: TF code for creating dynamic Resource Hierarchy folder structure

### PR DESCRIPTION
**Pull request by SudharsaneSivamany**
This Terraform code will help us to create dynamic resource hierarchy folder structure under the organization node in GCP.

### CODE REVIEW

**Comment by apeabody** [comment]
Hi @SudharsaneSivamany - Relevant output from the lint check:


```
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/examples/dynamic_example/README.md /tmp/tmp.Aa0XxfDG9M/generate_docs/workspace/examples/dynamic_example/README.md
6,29d5
< ## Requirements
< 
< | Name | Version |
< |------|---------|
< | <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 0.13 |
< | <a name="requirement_google"></a> [google](#requirement\_google) | >= 3.50 |
< 
< 
< ## Modules
< 
< | Name | Source | Version |
< |------|--------|---------|
< | <a name="module_folders"></a> [folders](#module\_folders) | ../../ | n/a |
< | <a name="module_sub_folders1"></a> [sub\_folders1](#module\_sub\_folders1) | ../../ | n/a |
< | <a name="module_sub_folders2"></a> [sub\_folders2](#module\_sub\_folders2) | ../../ | n/a |
< | <a name="module_sub_folders3"></a> [sub\_folders3](#module\_sub\_folders3) | ../../ | n/a |
< | <a name="module_sub_folders4"></a> [sub\_folders4](#module\_sub\_folders4) | ../../ | n/a |
< | <a name="module_sub_folders5"></a> [sub\_folders5](#module\_sub\_folders5) | ../../ | n/a |
< | <a name="module_sub_folders6"></a> [sub\_folders6](#module\_sub\_folders6) | ../../ | n/a |
< | <a name="module_sub_folders7"></a> [sub\_folders7](#module\_sub\_folders7) | ../../ | n/a |
< | <a name="module_sub_folders8"></a> [sub\_folders8](#module\_sub\_folders8) | ../../ | n/a |
< | <a name="module_sub_folders9"></a> [sub\_folders9](#module\_sub\_folders9) | ../../ | n/a |
< 
< 
34,35c10,11
< | <a name="input_folder_list"></a> [folder\_list](#input\_folder\_list) | Mention the folder structure | `any` | n/a | yes |
< | <a name="input_org_id"></a> [org\_id](#input\_org\_id) | The Organization Id | `string` | n/a | yes |
---
> | folder\_list | Mention the folder structure | `any` | n/a | yes |
> | org\_id | The Organization Id | `string` | n/a | yes |
41c17
< | <a name="output_folder_ids"></a> [folder\_ids](#output\_folder\_ids) | n/a |
---
> | folder\_ids | n/a |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
Checking for trailing whitespace
Checking for missing newline at end of file
Running shellcheck
Checking file headers
./examples/dynamic_example/outputs.tf
./examples/dynamic_example/versions.tf
./examples/dynamic_example/main.tf
./examples/dynamic_example/variables.tf
```

**Comment by SudharsaneSivamany** [comment]
> Hi @SudharsaneSivamany - Relevant output from the lint check:
> 
> ```
> Checking for documentation generation
> diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/examples/dynamic_example/README.md /tmp/tmp.Aa0XxfDG9M/generate_docs/workspace/examples/dynamic_example/README.md
> 6,29d5
> < ## Requirements
> < 
> < | Name | Version |
> < |------|---------|
> < | <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 0.13 |
> < | <a name="requirement_google"></a> [google](#requirement\_google) | >= 3.50 |
> < 
> < 
> < ## Modules
> < 
> < | Name | Source | Version |
> < |------|--------|---------|
> < | <a name="module_folders"></a> [folders](#module\_folders) | ../../ | n/a |
> < | <a name="module_sub_folders1"></a> [sub\_folders1](#module\_sub\_folders1) | ../../ | n/a |
> < | <a name="module_sub_folders2"></a> [sub\_folders2](#module\_sub\_folders2) | ../../ | n/a |
> < | <a name="module_sub_folders3"></a> [sub\_folders3](#module\_sub\_folders3) | ../../ | n/a |
> < | <a name="module_sub_folders4"></a> [sub\_folders4](#module\_sub\_folders4) | ../../ | n/a |
> < | <a name="module_sub_folders5"></a> [sub\_folders5](#module\_sub\_folders5) | ../../ | n/a |
> < | <a name="module_sub_folders6"></a> [sub\_folders6](#module\_sub\_folders6) | ../../ | n/a |
> < | <a name="module_sub_folders7"></a> [sub\_folders7](#module\_sub\_folders7) | ../../ | n/a |
> < | <a name="module_sub_folders8"></a> [sub\_folders8](#module\_sub\_folders8) | ../../ | n/a |
> < | <a name="module_sub_folders9"></a> [sub\_folders9](#module\_sub\_folders9) | ../../ | n/a |
> < 
> < 
> 34,35c10,11
> < | <a name="input_folder_list"></a> [folder\_list](#input\_folder\_list) | Mention the folder structure | `any` | n/a | yes |
> < | <a name="input_org_id"></a> [org\_id](#input\_org\_id) | The Organization Id | `string` | n/a | yes |
> ---
> > | folder\_list | Mention the folder structure | `any` | n/a | yes |
> > | org\_id | The Organization Id | `string` | n/a | yes |
> 41c17
> < | <a name="output_folder_ids"></a> [folder\_ids](#output\_folder\_ids) | n/a |
> ---
> > | folder\_ids | n/a |
> Error: Documentation generation has not been run, please run the
> 'make docker_generate_docs' command and commit the above changes.
> Checking for trailing whitespace
> Checking for missing newline at end of file
> Running shellcheck
> Checking file headers
> ./examples/dynamic_example/outputs.tf
> ./examples/dynamic_example/versions.tf
> ./examples/dynamic_example/main.tf
> ./examples/dynamic_example/variables.tf
> ```



> Hi @SudharsaneSivamany - Relevant output from the lint check:
> 
> ```
> Checking for documentation generation
> diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/examples/dynamic_example/README.md /tmp/tmp.Aa0XxfDG9M/generate_docs/workspace/examples/dynamic_example/README.md
> 6,29d5
> < ## Requirements
> < 
> < | Name | Version |
> < |------|---------|
> < | <a name="requirement_terraform"></a> [terraform](#requirement\_terraform) | >= 0.13 |
> < | <a name="requirement_google"></a> [google](#requirement\_google) | >= 3.50 |
> < 
> < 
> < ## Modules
> < 
> < | Name | Source | Version |
> < |------|--------|---------|
> < | <a name="module_folders"></a> [folders](#module\_folders) | ../../ | n/a |
> < | <a name="module_sub_folders1"></a> [sub\_folders1](#module\_sub\_folders1) | ../../ | n/a |
> < | <a name="module_sub_folders2"></a> [sub\_folders2](#module\_sub\_folders2) | ../../ | n/a |
> < | <a name="module_sub_folders3"></a> [sub\_folders3](#module\_sub\_folders3) | ../../ | n/a |
> < | <a name="module_sub_folders4"></a> [sub\_folders4](#module\_sub\_folders4) | ../../ | n/a |
> < | <a name="module_sub_folders5"></a> [sub\_folders5](#module\_sub\_folders5) | ../../ | n/a |
> < | <a name="module_sub_folders6"></a> [sub\_folders6](#module\_sub\_folders6) | ../../ | n/a |
> < | <a name="module_sub_folders7"></a> [sub\_folders7](#module\_sub\_folders7) | ../../ | n/a |
> < | <a name="module_sub_folders8"></a> [sub\_folders8](#module\_sub\_folders8) | ../../ | n/a |
> < | <a name="module_sub_folders9"></a> [sub\_folders9](#module\_sub\_folders9) | ../../ | n/a |
> < 
> < 
> 34,35c10,11
> < | <a name="input_folder_list"></a> [folder\_list](#input\_folder\_list) | Mention the folder structure | `any` | n/a | yes |
> < | <a name="input_org_id"></a> [org\_id](#input\_org\_id) | The Organization Id | `string` | n/a | yes |
> ---
> > | folder\_list | Mention the folder structure | `any` | n/a | yes |
> > | org\_id | The Organization Id | `string` | n/a | yes |
> 41c17
> < | <a name="output_folder_ids"></a> [folder\_ids](#output\_folder\_ids) | n/a |
> ---
> > | folder\_ids | n/a |
> Error: Documentation generation has not been run, please run the
> 'make docker_generate_docs' command and commit the above changes.
> Checking for trailing whitespace
> Checking for missing newline at end of file
> Running shellcheck
> Checking file headers
> ./examples/dynamic_example/outputs.tf
> ./examples/dynamic_example/versions.tf
> ./examples/dynamic_example/main.tf
> ./examples/dynamic_example/variables.tf
> ```

HI @apeabody , Fixed the lint issue.

File: examples/dynamic_example/main.tf [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+ * limitations under the License.
+ */
+
+locals {
+
+  sub_folders1_var = compact(flatten([for k, i in var.folder_list : i == [] ? [] : [for ip1, op1 in i : join("=1>", [k, ip1])]]))
+
+  sub_folders2_var = compact(flatten([for k, i in var.folder_list : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : join("=2>", [join("=1>", [k, ip1]), ip2])]]]))
+
+  sub_folders3_var = compact(flatten([for k, i in var.folder_list : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))
```
Comment (bharathkkb) [inline examples/dynamic_example/main.tf:23]: nit: can we just keep 3 levels and instruct users on how to do more levels if needed?

File: examples/dynamic_example/variables.tf [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+variable "org_id" {
+  type        = string
+  description = "The Organization Id "
+}
+
+variable "folder_list" {
+  type        = any
+  description = "Mention the folder structure"
```
Comment (bharathkkb) [inline examples/dynamic_example/variables.tf:24]:
```suggestion
  description = "Folder structure as a map"
```

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR @SudharsaneSivamany

File: examples/dynamic_example/main.tf (hunk 1 above)
Comment (SudharsaneSivamany) [reply examples/dynamic_example/main.tf:23]: Sure, Will keep till level 3.

**Comment by g-awmalik** [comment]
Lint errors:

```
Working in ./examples/dynamic_example ...
6 issue(s) found:

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 19:
  19:   sub_folders1_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : join("=1>", [k, ip1])]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 21:
  21:   sub_folders2_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : join("=2>", [join("=1>", [k, ip1]), ip2])]]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 21:
  21:   sub_folders2_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : join("=2>", [join("=1>", [k, ip1]), ip2])]]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 23:
  23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 23:
  23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)

  on main.tf line 23:
  23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md

tflint failed ./examples/dynamic_example 
```

**Comment by SudharsaneSivamany** [comment]
> Lint errors:
> 
> ```
> Working in ./examples/dynamic_example ...
> 6 issue(s) found:
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 19:
>   19:   sub_folders1_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : join("=1>", [k, ip1])]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 21:
>   21:   sub_folders2_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : join("=2>", [join("=1>", [k, ip1]), ip2])]]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 21:
>   21:   sub_folders2_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : join("=2>", [join("=1>", [k, ip1]), ip2])]]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 23:
>   23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 23:
>   23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> Warning: Comparing a collection with an empty list is invalid. To detect an empty collection, check its length. (terraform_empty_list_equality)
> 
>   on main.tf line 23:
>   23:   sub_folders3_var = compact(flatten([for k, i in var.folder_map : i == [] ? [] : [for ip1, op1 in i : op1 == [] ? [] : [for ip2, op2 in op1 : op2 == [] ? [] : [for ip3, op3 in op2 : join("=3>", [join("=2>", [join("=1>", [k, ip1]), ip2]), ip3])]]]]))
> 
> Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.1.1/docs/rules/terraform_empty_list_equality.md
> 
> tflint failed ./examples/dynamic_example 
> ```

Fixed the empty set lint issue.

**Review by bharathkkb — APPROVED** [review]

**Pull request merged**